lightning_client.go

```diff
@@ -3757,3 +3762,60 @@ func (s *lightningClient) SendCustomMessage(ctx context.Context,
 	_, err := s.client.SendCustomMessage(rpcCtx, rpcReq)
 	return err
 }
+
+// SubscribeCustomMessages subscribes to a stream of custom messages, optionally
+// filtering by peer and message type.
+func (s *lightningClient) SubscribeCustomMessages(ctx context.Context) (
+	chan<- CustomMessage, chan<- error, error) {
+
+	rpcCtx, cancel := context.WithTimeout(ctx, s.timeout)
+	defer cancel()
+
+	rpcCtx = s.adminMac.WithMacaroonAuth(rpcCtx)
+	rpcReq := &lnrpc.SubscribeCustomMessagesRequest{}
+
+	client, err := s.client.SubscribeCustomMessages(rpcCtx, rpcReq)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var (
+		// Buffer error channel by 1 so that consumer reading from this
+		// channel does not block our exit.
+		errChan = make(chan error, 1)
+		msgChan = make(chan CustomMessage)
+	)
+
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+
+		for {
+			msg, err := client.Recv()
+			if err != nil {
+				errChan <- fmt.Errorf("receive failed: %w", err)
```

Comment:
Instead of directly sending to the `errChan` we could use similar pattern as when we're writing to the `msgChan` to avoid a potential deadlock and a lingering goroutine.

```
sendErr := func(err error) {
		select {
		case errChan <- err:
		case <-ctx.Done():
		}
	}
```

Reply:
I don't think we're at risk of deadlock with `errChan` buffered by 1? Since this loop will only ever send 1 or 0 errors into it then exit?

Reply:
Yes indeed, missed that it's buffered.